src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +216,249 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
+            String n = intfc.getName();
+            int i = n.lastIndexOf('.');
+            String cn = i > 0 ? pn + "." + n.substring(i + 1) : pn + "." + n;
+            ClassDesc proxyDesc = ClassDesc.of(cn);
+            byte[] template = createTemplate(loader, proxyDesc, desc(intfc), uniqueName, methods);
+            var definer = new Lookup(intfc).makeHiddenClassDefiner(cn, template, Set.of(), DUMPER);
+
+            @SuppressWarnings("removal")
+            var sm = System.getSecurityManager();
+            Lookup lookup;
+            if (sm != null) {
+                @SuppressWarnings("removal")
+                var l = AccessController.doPrivileged((PrivilegedAction<Lookup>) () -> definer.defineClassAsLookup(true));
+                lookup = l;
+            } else {
+                lookup = definer.defineClassAsLookup(true);
+            }
+            WRAPPER_TYPES.add(lookup.lookupClass());
+            return new SoftReference<>(lookup);
+        }
+    };
+
+    private static Lookup getProxyClassLookup(Class<?> intfc) {
+        SoftReference<Lookup> r = PROXY_LOOKUPS.get(intfc);
+        if (r.refersTo(null)) {
+            PROXY_LOOKUPS.remove(intfc);
+            r = PROXY_LOOKUPS.get(intfc);
+        }
+        return r.get();
+    }
+
+    private static final List<ClassDesc> DEFAULT_RETHROWS = List.of(desc(RuntimeException.class), desc(Error.class));
+    private static final ClassDesc CD_UndeclaredThrowableException = desc(UndeclaredThrowableException.class);
+    private static final ClassDesc CD_IllegalAccessException = desc(IllegalAccessException.class);
+    private static final MethodTypeDesc MTD_void_Throwable = MethodTypeDesc.of(CD_void, CD_Throwable);
+    private static final MethodType MT_void_Lookup_MethodHandle_MethodHandle = methodType(void.class, Lookup.class, MethodHandle.class, MethodHandle.class);
+    private static final MethodType MT_Object_Lookup_MethodHandle_MethodHandle = MT_void_Lookup_MethodHandle_MethodHandle.changeReturnType(Object.class);
+    private static final MethodType MT_MethodHandle_Object = methodType(MethodHandle.class, Object.class);
+    private static final MethodTypeDesc MTD_void_Lookup_MethodHandle_MethodHandle = desc(MT_void_Lookup_MethodHandle_MethodHandle);
+    private static final MethodTypeDesc MTD_void_Lookup = MethodTypeDesc.of(CD_void, CD_MethodHandles_Lookup);
+    private static final MethodTypeDesc MTD_MethodHandle_MethodType = MethodTypeDesc.of(CD_MethodHandle, CD_MethodType);
+    private static final MethodTypeDesc MTD_Class = MethodTypeDesc.of(CD_Class);
+    private static final MethodTypeDesc MTD_int = MethodTypeDesc.of(CD_int);
+    private static final MethodTypeDesc MTD_String = MethodTypeDesc.of(CD_String);
+    private static final MethodTypeDesc MTD_void_String = MethodTypeDesc.of(CD_void, CD_String);
+    private static final String ORIGINAL_TARGET_NAME = "originalTarget";
+    private static final String ORIGINAL_TYPE_NAME = "originalType";
```

Comment:
Can drop "original" prefix?

```suggestion
    private static final String TARGET_NAME = "target";
    private static final String TYPE_NAME = "interfaceType";
```